### PR TITLE
Add a new 'item' field to TaskMap

### PR DESCRIPTION
**Pull request by uranusjr**
To support expanding tasks against a dict's *content* at runtime (e.g. when given `{"a": [1, 2]}`, expand against 1 and 2), the scheduler must be able to know how long each value is. This means we need to extend the TaskMap metadata store a bit, so it can not only store information on the dict itself, but also on the values, and at the same time.

A new field "item" is thus added. When it's set to an empty string (the default), the row stores information on the upstream's XCom push itself. Information about a dict's content is stored in separate rows, where "item" is set to the corresponding key (which must be str since that's required by Python's keyword argument unpacking).

First step toward #24021.

### CODE REVIEW

**Comment by uranusjr** [comment]
Ew, this faces the same problem as XCom—adding a new field to the primary key makes the length go over MySQL’s index limit.

**Comment by potiuk** [comment]
:(

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -2272,19 +2271,20 @@ def _record_task_map_for_downstreams(self, task: "Operator", value: Any, *, sess
...
-            raise UnmappableXComTypePushed(value)
-        task_map = TaskMap.from_task_instance_xcom(self, value)
-        max_map_length = conf.getint("core", "max_map_length", fallback=1024)
-        if task_map.length > max_map_length:
-            raise UnmappableXComLengthPushed(value, max_map_length)
-        session.merge(task_map)
+        session.merge(TaskMap.from_task_instance_xcom(self, value, item=""))
+        if isinstance(value, dict) and any(node.should_unpack_mapped_kwargs() for node in it2):
+            for task_map in TaskMap.from_unpacking_task_instance_xcom(self, value):
+                session.merge(task_map)
```
Comment (ashb) [inline airflow/models/taskinstance.py:2287]: Why do we create both `item=""` and item set to something here?

Comment (uranusjr) [reply airflow/models/taskinstance.py:2287]: We need the item set for the “drill-down” kind of task expansion (i.e. `expand_kwargs(xcom_arg)`, where we need to iterate against the content of `xcom_arg`). Technically we can avoid the `item=""` case for this particular case, but we still need it for something like

```python
t1 = ...  # XComArg to {"a": [1, 2]}

t2.expand_kwargs(t1)  # This expands against [1, 2].
t3.expand(x=t1)  # This expands against ["a"].
```

but I figured it’s easier to simply save oen additional row than doing another DAG traversal to figure out if there is a `t3`.

**Comment by uranusjr** [comment]
Thinking about this again, maybe we should not explode the dict. When we do

```python
xcom_arg  # Say this is {"a": [1, 2], "b": [3, 4, 5], "c": [6, 7]}
foo.expand_kwargs(xcom_arg)
```

we can only expand the task to either 2 or 3 instances (depending whether we implement the `zip` or `zip_longest` semantic). We don’t actually need three additional entries storing the length of each value, but only need to record the zipped value is 2 (or 3—I think we should store both values). This means we only need to add two additional column on TaskMap (without changing the primary key), `zip=2` and `zig_longest=3`. Both values would be NULL if the content is not kwarg-expandable.

**Comment by uranusjr** [comment]
Preferring the other approach.

**Pull request closed without merging**